networkit/cpp/io/NetworkitBinaryGraph.md

```diff
@@ -0,0 +1,57 @@
+NetworkitBinaryGraph format
+==========================
+This is a documentation of the NetworkitBinaryGraph file format. 
+The file starts with a header followed by the base data, adjacency lists and weights respectively. 
+The rest of this documentation describes the aforementioned blocks in detail.
+
+Header
+---------
+The file begins with a header that is structured as follows:
+```
+struct Header {
+
+    char magic[8];
+    uint64_t checksum;
+    uint64_t features; 
+    uint64_t nodes;
+    uint64_t chunks;
+    uint64_t offsetBaseData;
+    uint64_t offsetAdjLists;
+    uint64_t offsetWeights;
+};
+```
+- magic: A constant value used to identify the file format version.
+    - The current version is '*nkbg000*'
+- checksum: Currently not used
+- features: Contains the graph information bitwise
+    - Bit 0 : directed or undirected
+    - Bit 1-2 : the weight format of the graph. Below are the possible formats:
+        - 0 = no weights
+        - 1 = Weights are unsigned integers
+        - 2 = Weights are signed integers
+        - 3 = Weights are floats/doubles
+- nodes : The number of nodes the graph has
+- chunks: The number of chunks the nodes have been divided in
+- offsetBaseData: Offset of base data in the file 
+- offsetAdjLists: Offset of the adjacency lists in the file
+- offsetWeights: Offset of the weights in the file
+
+All offsets are relative to the beginning of the file.
+
+Base data
+------------
+```
+uint64_t sizeSum[nodes]: The sum of sizes of adjacency vertices with indices <= i
+uint64_t firstVertex[chunks]: The index of the first vertex of each chunks
```

Comment:
To simplify reading the file, it would be good to store one more element in both arrays that indicates the end (so you can e.g. use a single `memcpy` to copy the whole chunk).

Reply:
Can you elaborate on that? It's not clear to me what you would like to achieve.

Reply:
If you want to know how many vertices are in a chunk, with this structure you always need to check if the current chunk is the last chunk. If there was an additional value, for getting the size of chunk `i`, you could simply subtract `firstVertex[i]` from `firstVertex[i+1]`.

The same is true for `sizeSum`. The description "The sum of sizes of adjacency vertices with indices <= i" is also not clear. From the code I understand that this is a prefix sum of the node's degrees. Here, again, if I want to get the degree of node `i`, I need to check first if this is the first node, if yes, it is the degree, otherwise I need to subtract the previous entry from it. Here, an additional entry would only help if the definition was changed to be the exclusive prefix sum, i.e., if it was "indices < i".

Btw., it seems the number of edges is only written in the last entry of this array?

Reply:
Yes, the number of edges is only written in the last entry. We could also store it in the header if a simple way to access it is considered to be important.

For computing the size of a chunk: to make firstVertex[i + 1] - firstVertex[i] always valid, we would have to store a zero in the first index (and shift indices by 1). I think this is better handled in the reader. But you're right that it will reduce the number of branch mispredictions and is probably a worthwhile optimization.